### DISCUSSION TITLE
JSON string as template

### DISCUSSION THREAD

**Discussion by jimfum** [Handlebars-Net/Handlebars.Net#496] [Q&A]
Hi,

I'm trying to use the following JSON string as a template

```csharp
var template = @"{
  ""sentence1"" : ""Your custom text is: {{{customText}}}"",
  ...
}";
var f = handlebars.Compile(template);
f(new { customText = "quoted text follows - ""foo""" });
```
The render becomes an invalid JSON because of the un-escaped double quotes (as also highlighted below):
```json
{
  "sentence1" : "Your custom text is: quoted text follows - "foo"",
  ...
}";
```
Is there a way to solve it? Like provide custom text processor. I've already checked `ITextEncoder` but that's only used together with `{{` and not `{{{` as sort of expected. I need to execute my JSON encoding for every inserted text.

Thanks

**Comment by oformaniuk** [comment]
Hi @jimfum 
You may use [`HandlebarsTemplate<TextWriter, object, object>`](https://github.com/Handlebars-Net/Handlebars.Net/blob/558562b4bd015d0ae0d2ec8958351ea2c9d851a8/source/Handlebars/Handlebars.cs#L45) delegate to pass custom `TextWriter` and do text processing as you need.

**Reply by jimfum** [reply]
Thanks @zjklee. Not sure I follow so I'm thinking I haven't expressed myself well. The placeholder that handlebars fills in with text (`customText` in the example above) needs to pro post-processed (JSON escaped) by my custom code before it's embedded into the final rendered document. The above delegate gives me a TextWriter that will contain the final rendered document and through that I don't see how I'd do that.

**Reply by oformaniuk** [reply]
@jimfum 
In custom `TextWriter` implementation you may keep track for `"`. However, an implementation may be overcomplicated.

Do you need to stick to `{{{`? I assume you process JSON and do not really need HTML escaping. Custom implementation of `ITextEncoder` may solve your problem.

**Reply by jimfum** [reply]
The whole story is a bit more complicated. The JSON string represents a dictionary where the values are texts, possibly with some HTML syntax, that may contain also handlebars syntax. Eventually, the rendered values will be used within some bigger HTML, so keeping HTML escaping (i.e. respecting if the original text value contained `{{` or `{{{`)  is beneficial. 

The safe approach would be:
- Deserialize JSON into dictionary. Process (key, value) pairs one by one, handlebars compile the value, render and replace the original text with the rendered one.

That isn't very efficient. So I'm trying to have just one handlebars template, with JSON text, render once and then deserialize.

**Reply by oformaniuk** [reply]
@jimfum,

I'm still not sure I get it. Having miniaturized example in the form of test might help.

One other option I can think about is to declare a custom type to hold your `string` and define custom formatter. But this will not give you template-level control. In order to achieve template-level control you need to define custom helper that will ensure that output will be properly escaped.

**Reply by jimfum** [reply]
This may complete the picture. There's one main template (big 40kB HTML) with handlebars syntax defining the document structure that can get quite complex. There're many JSON files containing a dictionary `(key, text)` where the text also uses handlebars syntax and may contain bits of HTML. These JSONs contain many small texts.

* To get the final rendered document we combine the main template, one JSON file (selected based on certain criteria like spoken language) and the dynamic render input data.
* The main template references texts from the JSON though `{{key}}` or `{{{key}}}` (meaning we make the content of JSON to also become part of the handlebars context) based on the need, which is driven based on the JSON value containing additional HTML or not, and incoming render input data like `{{customerName}}` or `{{{customerName}}}`. 
* Having just the main template compiled isn't enough as that wouldn't process handlebars syntax in the JSON file. So the idea is to render the JSON file first (with the same handlebars context as the main template later) so we get the final texts there and then render the main template referencing those.

main template
```html
<html>
  <p>{{{sayHello}}}<p>
  <p>{{generatedOn}}</p>
</html>

```
version1.json
```json
{
"sayHello": "Welcome back, <b>{{customerName}}</b>"
}
```
version2.json
```json
{
"sayHello": "Hello <b>{{customerName}}</b>"
}
```

data for main template rendering:
```
{
  generatedOn = DateTime.Now,
  customerName = "John",
  +version(1|2).json deserialized dictionary content after handlebars rendering it from the original JSON
}
```

The problem is if the input data referenced from JSONs contain characters that need JSON encoding. E.g. `customerName = "O\"Neil"` (not a realistic example here). As mentioned above I'm aware it's solvable by not trying to optimize - compile & render every JSON value through handlebars one by one after it's deserialized.

**Reply by oformaniuk** [reply]
@jimfum , the thing is that I cannot get what manipulations with the json and dictionaries you're doing. Seeing it in the form of failing test helps to understand it better.
I understood that you're trying to optimize but I cannot understand hot template from json becomes part of the main template without rendering it first.

**Reply by rexm** [reply]
I don’t understand why these key-values that also contain Handlebars syntax aren’t helpers.

**Reply by jimfum** [reply]
@zjklee, exactly right, the template from json is rendered first, deserialized and used as an input to the main template. I'll add a test demonstrating the json manipulation. And to emphasize, I'm not pointing to any bug, it behaves as it should just not giving me what is needed.

@rexm, we use custom helpers quite a bit but don't see how they'd help here. The only idea coming to mind would be a helper, that does handlebars compile & render on the text it receives (the values from JSON) . That would generate a lot of memory allocation and GC as there're a lot of those to process. If your idea is different, could you pls demonstrate on the example above?

Thanks both of you.

**Reply by oformaniuk** [reply]
@jimfum, 
> I'll add a test demonstrating the json manipulation

Code (test in our case) is the best explanation of what you want to achieve. Try to expose what you want and what is failing as close to real scenario as possible but keeping only important details.

> And to emphasize, I'm not pointing to any bug, it behaves as it should just not giving me what is needed.

I understand, I'm just curious to make it work and do it in optimal way at the same time.

**Reply by jimfum** [reply]
I'm not sure myself now I can clearly demonstrate the challenges. Please find a couple of tests below. The reason for all of this is pretty much https://github.com/Handlebars-Net/Handlebars.Net/discussions/493.

```csharp
[Fact]
public void Test1()
{
	// {{{ here because we know the value from JSON contains HTML
	const string Template = @"<p>{{{dataFromJson.sayHello}}}</p><p>{{generatedOn}}</p>";
	// {{ would make more sense here but it's user's choice. neither option
	// should end up with exception
	const string Json1 = @"{
		""sayHello"": ""Hi <b>{{{customerName}}}</b>""
		// + many others
	}";

	var handlebars = Handlebars.Create();
	var renderMain = handlebars.Compile(Template);
	var renderJson = handlebars.Compile(Json1);

	dynamic data = new ExpandoObject();
	// the value isn't JSON escaped while we know it's going to be placed
	// into JSON string. issues expected.
	data.customerName = @"John "" Joshua";

	string renderedJson = renderJson(data);
	var jsonDeserialized = JsonConvert.DeserializeObject<Dictionary<string, string>>(renderedJson);

	data.generatedOn = "2022-01-01";
	data.dataFromJson = jsonDeserialized;

	var render = renderMain(data);

	// fails:
	// Newtonsoft.Json.JsonReaderException
	// After parsing a value an unexpected character was encountered:
	// J.Path 'sayHello', line 2, position 42.
	// EXPECTED as we produced invalid JSON string by injecting an unescaped substring (customerName)
	Assert.Equal(@"<p>Hi <b>John "" Joshua</b></p><p>2022-01-01</p>", render);
}

[Fact]
public void Test2()
{
	const string Template = @"<p>{{{dataFromJson.sayHello}}}</p><p>{{generatedOn}}</p>";
	// decided to use {{. makes more sense.
	const string Json1 = @"{
		""sayHello"": ""Hi <b>{{customerName}}</b>""
		// + many others
	}";

	var handlebars = Handlebars.Create();
	var renderMain = handlebars.Compile(Template);
	var renderJson = handlebars.Compile(Json1);

	dynamic data = new ExpandoObject();
	// JSON escaped ". Logical as we know we're injecting it into valid JSON string.
	data.customerName = @"John \"" Joshua";

	string renderedJson = renderJson(data);
	var jsonDeserialized = JsonConvert.DeserializeObject<Dictionary<string, string>>(renderedJson);

	data.generatedOn = "2022-01-01";
	data.dataFromJson = jsonDeserialized;

	var render = renderMain(data);

	// Newtonsoft.Json.JsonReaderException
	// Bad JSON escape sequence: \&.Path 'sayHello', line 2, position 42.
	// REASON: we JSON escaped customerName and then HTML escaped because of {{.
	// The right order would be HTML and then JSON espacing.
	Assert.Equal(@"<p>Hi <b>John &quot; Joshua</b></p><p>2022-01-01</p>", render);
}
```
